chain/index/events.go

```diff
@@ -280,7 +282,92 @@ func (si *SqliteIndexer) checkTipsetIndexedStatus(ctx context.Context, f *EventF
 	return ErrNotFound // Tipset is not indexed
 }
 
-// getTipsetKeyCidByHeight retrieves the tipset key CID for a given height.
+// checkRangeIndexedStatus verifies if a range of tipsets specified by the given height range is
+// indexed. It checks for the existence of non-null rounds at the range boundaries.
+func (si *SqliteIndexer) checkRangeIndexedStatus(ctx context.Context, minHeight abi.ChainEpoch, maxHeight abi.ChainEpoch) error {
+	head := si.cs.GetHeaviestTipSet()
+	if minHeight > head.Height() || maxHeight > head.Height() {
+		return ErrRangeInFuture
+	}
+
+	// Find the first non-null round in the range
+	startCid, startHeight, err := si.findFirstNonNullRound(ctx, minHeight, maxHeight)
+	if err != nil {
+		return xerrors.Errorf("failed to find first non-null round: %w", err)
+	}
+	// If all rounds are null, consider the range valid
+	if startCid == nil {
+		return nil
+	}
+
+	// Find the last non-null round in the range
+	endCid, endHeight, err := si.findLastNonNullRound(ctx, maxHeight, minHeight)
+	if err != nil {
+		return xerrors.Errorf("failed to find last non-null round: %w", err)
+	}
+	// If all rounds are null, consider the range valid
+	if endCid == nil {
+		return xerrors.Errorf("unexpected error finding last non-null round: all rounds are null but start round is not (%d to %d)", minHeight, maxHeight)
+	}
+
+	// Check indexing status for start and end tipsets
+	if err := si.checkTipsetIndexedStatus(ctx, startCid, startHeight); err != nil {
+		return err
+	}
+	if err := si.checkTipsetIndexedStatus(ctx, endCid, endHeight); err != nil {
+		return err
+	}
+	// Assume (not necessarily correctly, but likely) that all tipsets within the range are indexed
+
+	return nil
+}
+
+func (si *SqliteIndexer) checkTipsetIndexedStatus(ctx context.Context, tipsetKeyCid []byte, height abi.ChainEpoch) error {
+	exists, err := si.isTipsetIndexed(ctx, tipsetKeyCid)
+	if err != nil {
+		return xerrors.Errorf("failed to check if tipset at epoch %d is indexed: %w", height, err)
+	} else if exists {
+		return nil // has been indexed
+	}
+	return ErrNotFound
+}
+
+// findFirstNonNullRound finds the first non-null round starting from minHeight up to maxHeight.
+// It updates the minHeight to the found height and returns the tipset key CID.
+func (si *SqliteIndexer) findFirstNonNullRound(ctx context.Context, minHeight abi.ChainEpoch, maxHeight abi.ChainEpoch) ([]byte, abi.ChainEpoch, error) {
+	for height := minHeight; height <= maxHeight; height++ {
+		cid, err := si.getTipsetKeyCidByHeight(ctx, height)
+		if err != nil {
+			if !errors.Is(err, ErrNotFound) {
+				return nil, 0, xerrors.Errorf("failed to get tipset key cid for height %d: %w", height, err)
+			}
+			// else null round, keep searching
+			continue
+		}
+		minHeight = height // Update the minHeight to the found height
```

Comment:
why do we need to update `minHeight` to height here ? Can't we just return `height` ? IIUC, we're not passing a pointer to this function and so not really updating the original `minHeight` param

Reply:
Aha I see why. This is to zero down to range of non-null epochs. Yeah makes sense.

Reply:
No, you're right, this isn't necessary. It originally was in his first version because it was passed as a pointer that got updated but I suggested we not do that and this is an artefact of that change.